### DISCUSSION TITLE
Shorthand for string.IsNullOrWhiteSpace

### DISCUSSION THREAD

**Discussion by tolbxela** [dotnet/csharplang#6584] [Language Ideas]
I would really ❤️ to have a shorthand for `string.IsNullOrWhiteSpace`. 

It's the very nice feature I love to use on everyday basis. But it's also too long, verbose and very annoying to type in. 

Look at this, how often it's used in GutHub Repos!

![image](https://user-images.githubusercontent.com/4910589/197253630-5157b718-7c18-4a63-b279-452e0900654e.png)

We need something short, nice and cryptic 😉 as a shorthand. But I have no idea currently what from the !"§$%&/()=? to use.

Any Ideas?

**Comment by HaloFour** [comment]
`string.IsNullOrEmpty` has that beat by a huge margin:

<img width="703" alt="image" src="https://user-images.githubusercontent.com/10102290/197257009-9b2dc874-19c5-4662-9cd6-f90290fa8acd.png">

Either way, I don't think it's worth trying to come up with special glyphs to represent all of the various commonly-used helper methods that exist in the BCL as we'll quickly run out of glyphs long before we run out of methods.

**Reply by HaloFour** [reply]
You could always make it more concise via helper methods:

```cs
public static class StringExtensions {
    public static bool IsEmpty(this string text) => string.IsNullOrEmpty(text);
    public static bool IsWS(this string text) => string.IsNullOrWhitespace(text);
}

// usage
if (foo.IsEmpty()) {
    // ...
}
```

**Reply by tolbxela** [reply]
Yes, I know. But it's not cryptic enough! 😄

**Reply by tolbxela** [reply]
> `string.IsNullOrEmpty` has that beat by a huge margin:

Yes, I know.

I work a lot with FoxPro DBs (yep, it's still not dead) and use IsNullOrWhitespace() all the time, since all empty String values coming from FoxPro are padded with space. So I need to .Trim() everything or IsNullOrWhitespace() to check for empty values..

**Reply by tolbxela** [reply]
_special glyphs_  are cool! 😎 

They remember me of my early years of Perl programming. 😃

**Reply by HaloFour** [reply]
> I work a lot with FoxPro DBs (yep, it's still not dead) and use IsNullOrWhitespace() all the time, since all empty String values coming from FoxPro are padded with space.

I'd suggest that this is a very niche use case as that is a very uncommon behavior for databases, regardless of how dead they are (and I'd consider FoxPro/DB2 to be quite dead).

> They remember me of my early years of Perl programming. 😃

That ain't selling it.  😉

**Reply by RikkiGibson** [reply]
I once worked on an app which primarily worked with a database that used CHAR(n) fields instead of VARCHAR(n) fields. `string.Trim` was at the top of the perf trace. good times..

It might be useful if someone could recommend/link to best practices when working with such databases. I don't think a language-level shorthand for IsNullOrWhiteSpace is going to happen.

**Reply by tolbxela** [reply]
A language-level shorthand would be great to have independently from this special use case.

**Reply by CyrusNajmabadi** [reply]
I cannot envision us doing this.  Especially when the use case is motivated by a project targeting foxpro :)

**Reply by tolbxela** [reply]
It is not motivated by one project. I use this handy function a lot in my Web Apps, API and backend. Since it replaces three operations:
` ?.Trim().length > 0` and it's very cool. Such small things makes C# efficient and nice to use. 
I can compare C# with many other languages I have programmed already and I can definitely say, that programming C# is great! But still, it could be improved. 😃

**Reply by HaloFour** [reply]
It is useful.  So are thousands of other helper methods in the BCL.  That doesn't mean that they should be considered to be added as special syntax to the language.  It's already trivial to make a helper method to make this a lot terser, although, with autocomplete, I doubt it saves much in the way of keystrokes.

**Reply by LupusInferni315** [reply]
> It is not motivated by one project. I use this handy function a lot in my Web Apps, API and backend. Since it replaces three operations: ` ?.Trim().length > 0` and it's very cool. Such small things makes C# efficient and nice to use. I can compare C# with many other languages I have programmed already and I can definitely say, that programming C# is great! But still, it could be improved. 😃

but from now on, I'm liking the efficient nature of `?.Trim().Length > 0` it's a fair amount shorter than `string.IsNullOrWhitespace()`, perhaps having `IsNullOrWhitespace` is just unnecessary with that kind of shorthand.

**Comment by TahirAhmadov** [comment]
**Maybe** it can be added as a pattern? Like `str is not blank`, or `str is not space`. But I would still say it's not worth it; it would introduce a new way of doing something which has been done the old way for years. Basically, your evidence - the # of matches - is actually also an argument against changing it.

**Reply by RikkiGibson** [reply]
I would expect that if "active patterns" were added to the language then people would be able to create such a pattern themselves.

**Comment by ufcpp** [comment]
`is null or []`?

**Reply by jnm2** [reply]
That's IsNullOrEmpty. IsNullOrWhiteSpace would require allowing multiple `..` patterns inside list patterns 😄

```cs
bool IsNullOrWhiteSpace(string? str)
{
    return str is not [.., not (' ' or '\t' or '\v' or '\r' or '\n' /* etc */), ..];
}
```

**Comment by mattwar** [comment]
I propose we add an alternate method:  IsNullOrEmptyish.

**Reply by TahirAhmadov** [reply]
This opens a whole can of _wonders_. How about `char.ToSomeCase` and `Guid.AlmostZero`? We need to reimagine how we think about programming.

Still, all these change requests belong in the runtime repo. We have to follow the process.

**Comment by rbwhitaker** [comment]
Another workaround: an extension method.
```cs
public static bool IsNullOrWhitespace(this string value) => string.IsNullOrWhitespace(value);
```
Which shortens it a bit: `variable.IsNullOrWhitespace()` instead of `string.IsNullOrWhitespace(variable)`.

Not as short as a random glyph, but shorter anyway.

(Though I hate the aesthetics of a method that appears as an instance method but where it doesn't throw a `NullReferenceException` when invoked on `null`.)

**Reply by colejohnson66** [reply]
I personally never liked the idea that extension methods can take `null` without throwing. I get that your extension method is just fancy syntax for `StringExtensions.IsNullOrWhitespace(variable)`, but I don't like it.

**Comment by TonyValenti** [comment]
I have a "IsBlank" extension method for string? and love it.

**Reply by tolbxela** [reply]
Nice name for the method.

**Reply by WashingtonARamos** [reply]
I was about to comment about this, but as others have already pointed out, an extension method for it would be the better idea. I'd just name it "IsEmpty", as I see this a lot when doing validations with FluentValidation, it would make it very easy to understand for another dev working on the same project.

**Comment by MgSam** [comment]
Once they add emoji support into the language you could just do this:
``` C#
public class 😮 : Exception { }
public static bool 🤔(this string value) => string.IsNullOrWhitespace(value);

string s = null;
if (s.🤔()) 
     throw new 😮();
```

**Reply by HaloFour** [reply]
Silly, everyone know it's:

```cs
public class 😱 : Exception { }
```

**Reply by tolbxela** [reply]
Thanks for the great idea! But, unfortunately,  it doesn't work for me 😢 

![image](https://user-images.githubusercontent.com/4910589/197784716-dffda528-6eec-4ce3-bc50-726d61bfd619.png)

**Reply by HaloFour** [reply]
That it doesn't, because C# doesn't allow emoji in identifiers.

See: #3490

**Reply by tolbxela** [reply]
Sure... just wanted to make some fun of it 😸

**Reply by tolbxela** [reply]
My colleague have found a nice working OMG solution

```
public class 〵ꚝꯪꚝ〳: Exception{}
public class OMG : 〵ꚝꯪꚝ〳{}
```

I guess, he has definitely too much free time 😄

**Comment by LupusInferni315** [comment]
An alternative to this idea could be if string implemented an implicit conversion to bool, such that it would return the opposite of IsNullOrWhitespace, and could simply be called as such

```CSHARP
string variable = null; // or " ", or ""

if(!variable)
{
// do something here.
}
```

**Reply by HaloFour** [reply]
So this would be true?

```cs
if ("false") { ... }
```

**Reply by tolbxela** [reply]
a good question... 🤔

**Reply by LupusInferni315** [reply]
I suppose it would be

**Reply by rbwhitaker** [reply]
Indeed, you could embed entire programs in the language Whitespace and they'd all evaluate to `false`. I like that it subtly hints that programming in Whitespace is a "no" as far as C# is concerned. /sarcasm

**Reply by LupusInferni315** [reply]
I'll admit, that hadn't even crossed my mind, I entirely forgot that was even a thing, I only thought about the fact that I've never once used `string.IsNullOrEmpty`, I've always used `IsNullOrWhitespace`, as there's never been a time where I've needed to treat a whitespace only string as a not empty string.
